// web3-eth/lib/esm/utils/watch_transaction_for_confirmations.js
import { format } from 'web3-utils';
import { isNullish } from 'web3-validator';
import { TransactionMissingReceiptOrBlockHashError, TransactionReceiptMissingBlockNumberError, } from 'web3-errors';
import { transactionReceiptSchema } from '../schemas';
import { watchTransactionByPolling, } from './watch_transaction_by_pooling';
import { watchTransactionBySubscription } from './watch_transaction_by_subscription';
export function watchTransactionForConfirmations(web3Context, transactionPromiEvent, transactionReceipt, transactionHash, returnFormat) {
    if (isNullish(transactionReceipt) || isNullish(transactionReceipt.blockHash))
        throw new TransactionMissingReceiptOrBlockHashError({
            receipt: transactionReceipt,
            blockHash: format({ eth: 'bytes32' }, transactionReceipt === null || transactionReceipt === void 0 ? void 0 : transactionReceipt.blockHash, returnFormat),
            transactionHash: format({ eth: 'bytes32' }, transactionHash, returnFormat),
        });
    if (!transactionReceipt.blockNumber)
        throw new TransactionReceiptMissingBlockNumberError({ receipt: transactionReceipt });
    // As we have the receipt, it's the first confirmation that tx is accepted.
    transactionPromiEvent.emit('confirmation', {
        confirmations: format({ eth: 'uint' }, 1, returnFormat),
        receipt: format(transactionReceiptSchema, transactionReceipt, returnFormat),
        latestBlockHash: format({ eth: 'bytes32' }, transactionReceipt.blockHash, returnFormat),
    });
    // so a subscription for newBlockHeaders can be made instead of polling
    const provider = web3Context.requestManager.provider;
    if (provider.supportsSubscriptions()) {
        watchTransactionBySubscription({
            web3Context,
            transactionReceipt,
            transactionPromiEvent,
            returnFormat,
        });
    }
    else {
        watchTransactionByPolling({
            web3Context,
            transactionReceipt,
            transactionPromiEvent,
            returnFormat,
        });
    }
}
