// web3-eth-accounts/lib/commonjs/common/hardforks/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.hardforks = void 0;
/*
This file is part of web3.js.

web3.js is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

web3.js is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
const chainstart_json_1 = __importDefault(require("./chainstart.json"));
const dao_json_1 = __importDefault(require("./dao.json"));
const homestead_json_1 = __importDefault(require("./homestead.json"));
const tangerineWhistle_json_1 = __importDefault(require("./tangerineWhistle.json"));
const spuriousDragon_json_1 = __importDefault(require("./spuriousDragon.json"));
const byzantium_json_1 = __importDefault(require("./byzantium.json"));
const constantinople_json_1 = __importDefault(require("./constantinople.json"));
const petersburg_json_1 = __importDefault(require("./petersburg.json"));
const istanbul_json_1 = __importDefault(require("./istanbul.json"));
const muirGlacier_json_1 = __importDefault(require("./muirGlacier.json"));
const berlin_json_1 = __importDefault(require("./berlin.json"));
const london_json_1 = __importDefault(require("./london.json"));
const shanghai_json_1 = __importDefault(require("./shanghai.json"));
const arrowGlacier_json_1 = __importDefault(require("./arrowGlacier.json"));
const grayGlacier_json_1 = __importDefault(require("./grayGlacier.json"));
const mergeForkIdTransition_json_1 = __importDefault(require("./mergeForkIdTransition.json"));
const merge_json_1 = __importDefault(require("./merge.json"));
exports.hardforks = {
    chainstart: chainstart_json_1.default,
    homestead: homestead_json_1.default,
    dao: dao_json_1.default,
    tangerineWhistle: tangerineWhistle_json_1.default,
    spuriousDragon: spuriousDragon_json_1.default,
    byzantium: byzantium_json_1.default,
    constantinople: constantinople_json_1.default,
    petersburg: petersburg_json_1.default,
    istanbul: istanbul_json_1.default,
    muirGlacier: muirGlacier_json_1.default,
    berlin: berlin_json_1.default,
    london: london_json_1.default,
    shanghai: shanghai_json_1.default,
    arrowGlacier: arrowGlacier_json_1.default,
    grayGlacier: grayGlacier_json_1.default,
    mergeForkIdTransition: mergeForkIdTransition_json_1.default,
    merge: merge_json_1.default,
};
