// web3-validator/lib/esm/keywords/eth.js
/*
This file is part of web3.js.

web3.js is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

web3.js is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
import { Web3ValidatorError } from '../errors';
import { parseBaseType } from '../utils';
import { isAddress, isBlockNumber, isBlockNumberOrTag, isBloom, isBoolean, isBytes, isFilterObject, isHexStrict, isInt, isNumber, isString, isUInt, } from '../validation';
const createErrorObject = (message, value) => ({
    message,
    keyword: 'eth',
    params: { value },
});
export const metaSchema = {
    title: 'Web3 Ethereum Compatible Types',
    type: 'string',
};
const extraTypes = ['hex', 'number', 'blockNumber', 'blockNumberOrTag', 'filter', 'bloom'];
const compile = (type, parentSchema, it) => {
    var _a, _b;
    const typePropertyPresent = Object.keys(parentSchema).includes('type');
    if (typePropertyPresent) {
        throw new Web3ValidatorError([
            {
                keyword: 'eth',
                message: 'Either "eth" or "type" can be presented in schema',
                params: { eth: type },
                instancePath: '',
                schemaPath: (_a = it.schemaPath.str) !== null && _a !== void 0 ? _a : '',
            },
        ]);
    }
    const { baseType } = parseBaseType(type);
    if (!baseType && !extraTypes.includes(type)) {
        throw new Web3ValidatorError([
            {
                keyword: 'eth',
                message: `Eth data type "${type}" is not valid`,
                params: { eth: type },
                instancePath: '',
                schemaPath: (_b = it.schemaPath.str) !== null && _b !== void 0 ? _b : '',
            },
        ]);
    }
    const validate = (data) => {
        let result = false;
        if (baseType) {
            // eslint-disable-next-line default-case
            switch (baseType) {
                case 'bool':
                    result = isBoolean(data);
                    break;
                case 'bytes':
                    result = isBytes(data, { abiType: type });
                    break;
                case 'string':
                    result = isString(data);
                    break;
                case 'uint':
                    result = isUInt(data, { abiType: type });
                    break;
                case 'int':
                    result = isInt(data, { abiType: type });
                    break;
                case 'address':
                    result = isAddress(data);
                    break;
                case 'tuple': {
                    throw new Error('"tuple" type is not implemented directly.');
                }
            }
        }
        else {
            switch (type) {
                case 'hex':
                    result = isHexStrict(data);
                    break;
                case 'number':
                    result = isNumber(data);
                    break;
                case 'blockNumber':
                    result = isBlockNumber(data);
                    break;
                case 'blockNumberOrTag':
                    result = isBlockNumberOrTag(data);
                    break;
                case 'filter':
                    result = isFilterObject(data);
                    break;
                case 'bloom':
                    result = isBloom(data);
                    break;
                default:
                    validate.errors = [createErrorObject(`can not identity "${type}"`, data)];
                    return false;
            }
        }
        if (!result) {
            validate.errors = [createErrorObject(`must pass "${type}" validation`, data)];
        }
        return result;
    };
    return validate;
};
export const ethKeyword = {
    keyword: 'eth',
    compile,
    errors: true,
    modifying: false,
    metaSchema,
};
